ducktape/mark/_mark.py

```diff
@@ -0,0 +1,359 @@
+            },
+            {
+                "x": 2,
+                "y": "a"
+            },
+            {
+                "x": 2,
+                "y", "b"
+            }
+        ]
+    """
+    # Establish an ordering of the keys
+    key_list = [k for k in d.keys()]
+
+    expanded = []
+    values_list = [d[k] for k in key_list]  # list of lists
+    for v in itertools.product(*values_list):
+        # Iterate through the cartesian product of the lists of values
+        # One dictionary per element in this cartesian product
+        new_dict = {}
+        for i in range(len(key_list)):
+            new_dict[key_list[i]] = v[i]
+        expanded.append(new_dict)
+    return expanded
+
+
+def matrix(**kwargs):
+    """Function decorator used to parametrize with a matrix of values.
+    Decorating a function or method with @matrix marks it with the Matrix mark. When expanded using the
+    MarkedFunctionExpander, it yields a list of TestContext objects, one for every possible combination
+    of arguments.
+
+    Example::
+
+        @matrix(x=[1, 2], y=[-1, -2])
+        def g(x, y):
+            print "x = %s, y = %s" % (x, y)
+
+        for ctx in MarkedFunctionExpander(..., function=g, ...).expand():
+            ctx.function()
+
+        # output:
+        # x = 1, y = -1
+        # x = 1, y = -2
+        # x = 2, y = -1
+        # x = 2, y = -2
+    """
+    def parametrizer(f):
+        Mark.mark(f, Matrix(**kwargs))
+        return f
+    return parametrizer
+
+
+def parametrize(**kwargs):
+    """Function decorator used to parametrize its arguments.
+    Decorating a function or method with @parametrize marks it with the Parametrize mark.
+
+    Example::
+
+        @parametrize(x=1, y=2 z=-1)
+        @parametrize(x=3, y=4, z=5)
+        def g(x, y, z):
+            print "x = %s, y = %s, z = %s" % (x, y, z)
+
+        for ctx in MarkedFunctionExpander(..., function=g, ...).expand():
+            ctx.function()
+
+        # output:
+        # x = 1, y = 2, z = -1
+        # x = 3, y = 4, z = 5
+    """
+    def parametrizer(f):
+        Mark.mark(f, Parametrize(**kwargs))
+        return f
+    return parametrizer
+
+
+def ignore(*args, **kwargs):
+    """
+    Test method decorator which signals to the test runner to ignore a given test.
+
+    Example::
+
+        When no parameters are provided to the @ignore decorator, ignore all parametrizations of the test function
+
+        @ignore  # Ignore all parametrizations
+        @parametrize(x=1, y=0)
+        @parametrize(x=2, y=3)
+        def the_test(...):
+            ...
+
+    Example::
+
+        If parameters are supplied to the @ignore decorator, only ignore the parametrization with matching parameter(s)
+
+        @ignore(x=2, y=3)
+        @parametrize(x=1, y=0)  # This test will run as usual
+        @parametrize(x=2, y=3)  # This test will be ignored
+        def the_test(...):
+            ...
+    """
+    if len(args) == 1 and len(kwargs) == 0:
```

Comment:
Isn't the wrapped function always going to be the first arg, and must be there anyway? So maybe this can be made clearer if the method is defined as

```
def ignore(f, *args, **kwargs)
```

Additionally, do we want the `*args` part, or should we only have `**kwargs`? It looks like the positional arguments are ignored below (the `ignorer` nested function doesn't use them), but by including `*args`, we allow extra positional arguments to be provided.